Adding press-enter options for VB

jcouv opened the pull request:
@CyrusNajmabadi @Pilchie  for early review.

A few open issues:
- making a similar change to C# (will be a bit trickier because the setting needs to be backwards compatible)
- I noticed that the equivalent C# setting appears in test automation, should I put some test automation for the VB setting too?
- although it's not related to the current PR, I noticed that `VisualBasicCompletionService.GetRules` is called on every key press, can you confirm that is acceptable?

CyrusNajmabadi commented:
> I noticed that VisualBasicCompletionService.GetRules is called on every key press, can you confirm that is acceptable?

Do you see the same for CSharpCompletionService.GetRules?

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionOptions.vb:
```diff
@@ -0,0 +1,11 @@
+    Friend Class VisualBasicCompletionOptions
+        Public Const FeatureName As String = "VisualBasic Completion"
+
+        Public Shared AddNewLineOnEnterAfterFullyTypedWord As Options.Option(Of Integer) = New Options.Option(Of Integer)(FeatureName, "Add New Line On Enter After Fully Typed Word", defaultValue:=EnterKeyRule.Always)
```
Simplify to Option(of integer).

CyrusNajmabadi replied:
Wrap onto two lines.  If i can't see the whole line in github, it's too long.

CyrusNajmabadi replied:
use nameof(AddNewLineOnEnterAfterFullyTypedWord) instead of the string you created.

CyrusNajmabadi replied:
Not a fan of having a new option.  Would prefer to have a single PerLanguageOption here for C# and VB to share.

jcouv replied:
Simplified `Options.Option` to `[Option]` and split the line.
I'll look into `PerLanguageOption`. What's the advantage?

CyrusNajmabadi replied:
The advantage is that we can now use the same option across langauges (including TypeScript).  We can then move the chekcing of that option to a more base layer (which we probably should be doing as well), and all Roslyn languages will pick up support for it for free.

CyrusNajmabadi replied:
oh darn.  i forgot Option was a VB keyword.   in that case you can pick either style :)

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionOptionsProvider.vb:
```diff
@@ -0,0 +1,19 @@
+    Friend Class VisualBasicCompletionOptionsProvider
+        Implements IOptionProvider
+
+        Dim _options As IEnumerable(Of IOption) = ImmutableArray.Create(DirectCast(VisualBasicCompletionOptions.AddNewLineOnEnterAfterFullyTypedWord, IOption))
```
easier is to just say ```ImmutableArray.Create(of IOption)(...)```

jcouv replied:
Thanks. Fixed

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionService.vb:
```diff
@@ -57,15 +57,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion
-                defaultCommitCharacters:=CompletionRules.Default.DefaultCommitCharacters,
-                defaultEnterKeyRule:=EnterKeyRule.Always)
-
         Public Overrides Function GetRules() As CompletionRules
```
Can you follow the same pattern we see in CSharpCompletionService?  i.e. with the interlocking and whatnot.  That will help avoid allocations here.

jcouv replied:
Done.

jcouv commented:
Yes, I see the same behavior with `GetRules` in C#.

CyrusNajmabadi commented on src/VisualStudio/VisualBasic/Impl/BasicVSResources.Designer.vb:
```diff
@@ -110,6 +110,15 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic
+        '''</summary>
+        Friend Shared ReadOnly Property Option_AlwaysInsertNewlineOnEnter() As String
+            Get
+                Return ResourceManager.GetString("Option_AlwaysInsertNewlineOnEnter", resourceCulture)
```
my preference is to have tehse exactly match the english string, just with underscores.  So if the english string is "Always insert newline on enter" then the resource name should be Option_Always_insert_newline_on_enter"

jcouv replied:
Done

CyrusNajmabadi commented on src/VisualStudio/VisualBasic/Impl/BasicVSResources.resx:
```diff
@@ -246,4 +246,16 @@
   <data name="Option_CompletionLists" xml:space="preserve">
     <value>Completion Lists</value>
   </data>
+  <data name="Option_SelectionInCompletionList" xml:space="preserve">
```
What ist his for?

jcouv replied:
The string for the group of radio buttons.

CyrusNajmabadi commented on src/VisualStudio/VisualBasic/Impl/Options/IntelliSenseOptionPageControl.xaml:
```diff
@@ -15,14 +15,20 @@
-
+                </StackPanel>
+            </GroupBox>
+            <GroupBox x:Uid="SelectionInCompletionListGroupBox"
```
I'm not sure this warrants a whole new group.  Can we just have a label or something and then the radios below it?

CyrusNajmabadi replied:
i.e. more like this:

![image](https://cloud.githubusercontent.com/assets/4564579/15801170/907a4da4-2a41-11e6-99b6-e1151dc039b6.png)

CyrusNajmabadi commented on src/VisualStudio/VisualBasic/Impl/Options/IntelliSenseOptionPageControl.xaml:
```diff
@@ -15,14 +15,20 @@
+            <GroupBox x:Uid="SelectionInCompletionListGroupBox"
+                     Header="{x:Static local:IntelliSenseOptionPageStrings.Option_SelectionInCompletionList}">
+                <StackPanel>
+                    <RadioButton  GroupName="InsertNewlineOnEnterRadio" x:Name="Never_insert_new_line" Content="{x:Static local:IntelliSenseOptionPageStrings.Option_NeverInsertNewlineOnEnter}"/>
```
inconsistent names.  in one place you have Never_insert_new_line and right after that you have Option_NeverInsertNewlineOnEnter.  These names should be as consistent as possible across all the places you have to touch.

jcouv replied:
Fixed.

jcouv replied:
Fixed.

CyrusNajmabadi commented:
Ah, i see the problem you're running into.    So here's my recommendation.  We introduce a new PerLanguageOption at the language neutral layer for this.

C# will have to still keep around it's existing CSharpCompletionOptions.AddNewLineOnEnterAfterFullyTypedWord option for back compat.

Then, inside the completion service, you see if your new PerLangaugeOption is set to something other than the default value.  If so, you return that.  If not, you fallback to checking the old option.

--

Alternatively, you get rid of the CSharp option.  But you'll need to fix up the CSharp settigns serializer to bridge between that old option and hte new one.

CyrusNajmabadi commented:
I think the latter approach actually sounds the best to me.  Just have one per-language option.  But update our serializer to properly bridge between the old value and hte new.

jcouv commented:
Here's the latest looks:
![image](https://cloud.githubusercontent.com/assets/12466233/15801424/e64b41aa-2a48-11e6-9db5-e883b1078a04.png)

jcouv commented:
Pushed a change that addresses the feedback, except the per-language-setting stuff (will come later).

CyrusNajmabadi commented:
Looks good.  Just needs work on the spacing.

jcouv commented:
Here's the latest screenshot (with improved spacing), in the context of C# options:
![image](https://cloud.githubusercontent.com/assets/12466233/15835096/54992adc-2be3-11e6-88f6-5244cd696b9b.png)

CyrusNajmabadi commented:
nice.  much better.

jcouv commented:
Pushed a change with per-language setting as we discussed. I updated some tests, but feel free to suggest if I need to add new tests.

Two notable things:
1. I removed a dead line of code that was getting completion rules in `GetCompletionsAsync`
2. Completions that come from the GAC use their own completion rule (see `AbstractReferenceDirectiveCompletionProvider`). I added a note in one of the tests that surprised me.

Pilchie commented:
Tagging @rchande and the rest of @dotnet/roslyn-ide for review too.

cston commented:
@dotnet-bot retest windows_debug_uint32_prtest please
@dotnet-bot retest windows_debug_uint64_prtest please

cston commented:
@dotnet-bot retest windows_debug_uint32_prtest please

jcouv commented:
@dotnet-bot retest windows_debug_unit32_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/windows_debug_unit32_prtest/316/

jcouv commented:
@dotnet-bot retest windows_debug_unit32_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/windows_debug_unit32_prtest/333/
// Retest reason: timed out

jcouv commented:
@dotnet-bot retest windows_debug_unit64_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/windows_debug_unit64_prtest/314/
// Retest reason:

jcouv commented:
@rchande @dotnet/roslyn-ide Kindly review and provide feedback.

jcouv commented:
@rchande This is the PR I mentioned yesterday. Please review.

rchande commented:
:+1:

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/CompletionOptions.cs:
```diff
@@ -14,6 +14,9 @@ internal static class CompletionOptions
         public static readonly PerLanguageOption<bool> TriggerOnTyping = new PerLanguageOption<bool>(FeatureName, "TriggerOnTyping", defaultValue: true);
         public static readonly PerLanguageOption<bool> TriggerOnTypingLetters = new PerLanguageOption<bool>(FeatureName, "TriggerOnTypingLetters", defaultValue: true);
 
+        public static readonly PerLanguageOption<int> AddNewLineOnEnterAfterFullyTypedWord =
```
Can this be a ```PerLanguageOption<EnterKeyRule>``` instead?  then you'd need less int casting all over the place.

CyrusNajmabadi replied:
This is not the right name for hte option.  It should be something like EnterKeyBehavior

jcouv replied:
Good idea on using `EnterKeyRule` instead of `int`.
For the name of the setting, I still need it serialized under the same name.

How about this?
```C#
public static readonly PerLanguageOption<EnterKeyRule> EnterKeyBehavior =
            new PerLanguageOption<EnterKeyRule>(FeatureName, "AddNewLineOnEnterAfterFullyTypedWord", defaultValue: EnterKeyRule.Default);
```

CyrusNajmabadi replied:
Hrmm... i'm not certain what i think is best.  Part of me things we should ahve this new otion, which serializes under the name "EnterKeyBehavior".  We then keep the old option around and migrate it forward if we see that it is there.    We put an Obsolete attribute on the old one to tell people to not use it anymore though.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/CSharpSettingsManagerOptionSerializer.cs:
```diff
@@ -373,9 +399,33 @@ public override bool TryPersist(OptionKey optionKey, object value)
+        /// </summary>
+        private int ConvertFromEnterKeyRule(object value)
+        {
+            EnterKeyRule rule = (EnterKeyRule)value;
```
var.

jcouv replied:
Done

CyrusNajmabadi commented:
LGTM if you can address the issues i pointed out.

CyrusNajmabadi commented:
What's the cause of the unit test failure.

jcouv commented:
I do not know why CI failed. The log doesn't show any specific unittest failing ("xunit produced no error output but had exit code 1").

jcouv commented:
Latest commit fixes the option name, uses `EnterKeyRule` instead of `int`, and upgrades old option when the new option is not present.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CSharpCompletionService.cs:
```diff
@@ -78,9 +78,12 @@ public override CompletionRules GetRules()
-                : EnterKeyRule.Never;
+            var rule = options.GetOption(CompletionOptions.EnterKeyBehavior, LanguageNames.CSharp);
+
+            if (rule == EnterKeyRule.Default)
```
Comment explaining this.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionService.vb:
```diff
@@ -57,17 +57,29 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion
+
+            Dim rule = options.GetOption(CompletionOptions.EnterKeyBehavior, LanguageNames.VisualBasic)
+
+            If rule = EnterKeyRule.Default Then
```
Comment explaining the historical nature of this.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/CSharpSettingsManagerOptionSerializer.cs:
```diff
@@ -276,6 +281,44 @@ private bool FetchStyleBool(string settingName, out object value)
+            }
+
+            // if the EnterKeyBehavior setting cannot be loaded, then attempt to load and upgrade the legacy AddNewLineOnEnterAfterFullyTypedWord setting
+            if (base.TryFetch(CSharpCompletionOptions.AddNewLineOnEnterAfterFullyTypedWord, out value))
```
nice.

CyrusNajmabadi commented:
:+1:

CyrusNajmabadi commented:
great comment. can't wait to have this in.  Thanks Julien!

jcouv commented:
Cool. Thanks for all the pointers and feedback!

CyrusNajmabadi commented:
retest vsi please
